docs/2024/puzzles/day15.md

```diff
@@ -2,10 +2,402 @@ import Solver from "../../../../../website/src/components/Solver.js"
+      .flatMap(_ match
+        case WarehouseCell.Empty => Some(0)
+        case WarehouseCell.Wall => None
+        case WarehouseCell.Box =>
+          getStackLengthFromTowards(nextRow, nextCol, d).map(_ + 1)
+        case WarehouseCell.Robot =>
+          throw AssertionError("should not reach robot in traversal"))
+
+  def move(d: Direction): Warehouse =
+    getStackLengthFromTowards(robotRow, robotCol, d) match
+      case None => this
+      case Some(stackLength) =>
+        val (newRobotRow, newRobotCol) = d(robotRow, robotCol)
+        val newCells = cells.updated(Seq(
+          ((robotRow, robotCol), WarehouseCell.Empty),
+          ((newRobotRow, newRobotCol), WarehouseCell.Robot)) ++
+          (if stackLength > 0
+           // stackLength + 1 because we want the cell after the stack
+           then Seq((d(robotRow, robotCol, stackLength + 1), WarehouseCell.Box))
+           else Seq(/* only the robot moves */)))
+        Warehouse(newCells, newRobotRow, newRobotCol)
+```
+
+Some nifty Scala features at play:
+* `.lift` lets us turn a potentially exception-throwing method (`ArraySeq.apply` out of bounds) into one that returns an `Option`, which we can further `flatMap` with the `Option`-valued wall/box/empty logic.
+* We define an `apply` method on `Direction` so that we can compute the updated position for `(row, col)` in direction `d` simply as `d(row, col)` (or optionally repeatedly, `d(row, col, n)`).
+
+  ```scala
+  enum Direction:
+    /* ... */
+    def apply(row: Int, col: Int, n: Int = 1): (Int, Int) =
+      this match
+        case Up => (row - n, col)
+        case Down => (row + n, col)
+        case Left => (row, col - n)
+        case Right => (row, col + n)
+  ```
+* `ArraySeq`s (like other immutable collections) provide an `updated` method that produces a new `ArraySeq` with just one element changed. That is sufficient for our needs, but we can make it even easier for ourselves with our own `updated` that does the same but for multiple elements at once, and with 2D indices rather than 2 levels of nesting. This is the only place we will make use of mutation in this program and we will be very careful about it.
+
+  ```scala
+  extension (cells: ArraySeq[ArraySeq[WarehouseCell]])
+                               // ((row, col), new cell)
+    def updated(updates: Iterable[((Int, Int), WarehouseCell)])
+        : ArraySeq[ArraySeq[WarehouseCell]] =
+      // Construct a fresh array (*not* using the one backing the ArraySeq),
+      val newCells = cells.map(_.toArray).toArray
+                                                  // mutate it,
+      updates.foreach{ case ((row, col), cell) => newCells(row)(col) = cell }
+      // and construct a new ArraySeq backed by it. Possibly unsafe if the
+      // backing array remains accessible outside of the ArraySeq, but safe
+      // here because the method only returns the ArraySeq.
+      ArraySeq.unsafeWrapArray(newCells).map(ArraySeq.unsafeWrapArray)
```

Comment:
Minor, might be able to avoid these unsafes by using `Seq` everywhere instead of the `ArraySeq` implementation type, and using `.toSeq` instead of the `.toArray`s. `Seq` is still immutable, so it won't compromise your morals!

Reply:
Ah but I know the size ahead of time and I know I will be doing a lot of random accesses. `Seq` doesn’t guarantee anything about that, right? (I’ll add this to my reasoning note)

Reply:
Yeah, but the typical implementation of Seq is the Vector, which offers basically O(1) everything. See [docs](https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html).

Reply:
My personal approach is to not rely on what the typical/default implementation is, but rather to only assume what the docs guarantee, and `Seq` and `IndexedSeq` would not quite guarantee enough. I would also have thought that `Vector`s provide O(1) random access but the [Scala 3 docs say](https://www.scala-lang.org/api/3.3.1/scala/collection/immutable/Vector.html#) it's O(log n).

Reply:
(The [collections guide](https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html) calls that "effectively constant" since the base of the log is quite high, so that it can never take more than a small number of steps.)